r_torch_tutorial.Rmd

```diff
@@ -0,0 +1,191 @@
+
+### Convolution Neural Network
+As is mentioned in the previous paragraph, a deep learning network is consisted of multiple blocks. For image classification task the most frequently used block is convolution neural network (CNN). Suppose we have a input of size $(N, C_{\mbox{in}}, H, W)$ and weight $W$ of size $(C_{\mbox{out}}, C_{\mbox{in}}, k, k)$ with stride $s$ and padding $p$, the output size will be $(N, C_{\mbox{out}}, H_{\mbox{out}}, W_{\mbox{out}})$ where
+$$
+  H_{\mbox{out}} = \left\lfloor\frac{H + 2p - k}{s}\right\rfloor + 1 \\
+  W_{\mbox{out}} = \left\lfloor\frac{W + 2p - k}{s}\right\rfloor + 1
+$$
+and 
+$$
+  \mbox{out}[n, j, x, y] = \mbox{bias}_j + \mbox{inner_product}(W[j,:,:,:], \mbox{in}[n, :, 1+(x-1)s:1+(x-1)s+k, 1+(y-1)s:1+(y-1)s+k])
+$$
+$i.e.$ each output value is the inner product result of two 4-d array of size $[1,C_\mbox{in}, k, k]$ plus bias. This block is widely used in image tasks has achieved good result and is the basic unit in foundation networks Alexnet[4] and Resnet[5].
+
+
+## Implementation
+### Installation
+To install torch in Rstudio, we should first install `torch` package. In our example, we also need `luz` and `torchvision`, therefore we run
+```{r Install, eval=FALSE}
+install.packages("torch")
+install.packages("torchvision")
+install.packages("luz")
+``` 
+and then run
+```{r Torch Installation, eval=FALSE}
+library(torch)
+library(torchvision)
+library(luz)
+```
+
+### Fetching Dataset
+In this example, we would like to identify the digits in the famous hand-written digit dataset `MNIST` using pure convolution neural network.
+```{r Dataset}
+dir <- "./dataset/mnist"
+
+train_ds <- mnist_dataset(
+  dir,
+  download = TRUE,
+  transform = transform_to_tensor
+)
+
+test_ds <- mnist_dataset(
+  dir,
+  train = FALSE,
+  transform = transform_to_tensor
+)
+
+train_dl <- dataloader(train_ds, batch_size = 128, shuffle = TRUE)
+test_dl <- dataloader(test_ds, batch_size = 128)
+```
+
+The `batch_size` parameter is the number of inputs that are processed in parallel and its value is decided by the computation capacity of the processor. An example of the input is shown as follow:
+```{r input, fig.height=3, fig.width=3}
+image <- train_ds$data[1,1:28,1:28]
+image_df <- melt(image)
+ggplot(image_df, aes(x=Var2, y=Var1, fill=value))+
+  geom_tile(show.legend = FALSE) + 
+  xlab("") + ylab("") +
+  scale_fill_gradient(low="white", high="black")
+```
+
+### Building up the network
+```{r Network}
+net <- nn_module(
+  "Net",
+  
+  initialize = function() {
+    self$conv1 <- nn_conv2d(1, 32, 3, 1)
+    self$conv2 <- nn_conv2d(32, 64, 3, 1)
+    self$dropout1 <- nn_dropout2d(0.25)
+    self$dropout2 <- nn_dropout2d(0.5)
+    self$fc1 <- nn_linear(9216, 128)
+    self$fc2 <- nn_linear(128, 10)
+  },
+  
+  forward = function(x) {
+    x %>%                                  # N * 1 * 28 * 28
```

Comment:
There should be no additional single hashtag headers in the chapter. (If there are, new chapters will be created.)